aws/resource_aws_workspace.go

```diff
@@ -0,0 +1,390 @@
+			Refresh:      workspaceRefreshStateFunc(conn, d.Id()),
+			PollInterval: 30 * time.Second,
+			Timeout:      10 * time.Minute,
+		}
+
+		_, err = stateConf.WaitForState()
+		if err != nil {
+			return fmt.Errorf("workspace %q properties was not modified: %s", d.Id(), err)
+		}
+		log.Printf("[DEBUG] Workspace %q properties are modified", d.Id())
+		d.SetPartial("workspace_properties")
+	}
+
+	if d.HasChange("tags") {
+		log.Printf("[DEBUG] Modifying workspace %q tags...", d.Id())
+		o, n := d.GetChange("tags")
+		if err := WorkspacesUpdateTags(conn, d.Id(), o, n); err != nil {
+			return fmt.Errorf("error updating workspaces %q tags: %s", d.Id(), err)
+		}
+		log.Printf("[DEBUG] Workspace %q tags are modified", d.Id())
+		d.SetPartial("tags")
+	}
+
+	d.Partial(false)
+
+	return resourceAwsWorkspaceRead(d, meta)
+}
+
+func resourceAwsWorkspaceDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	err := workspaceDelete(d.Id(), conn)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func workspaceDelete(id string, conn *workspaces.WorkSpaces) error {
+	log.Printf("[DEBUG] Terminating workspace %q", id)
+	_, err := conn.TerminateWorkspaces(&workspaces.TerminateWorkspacesInput{
+		TerminateWorkspaceRequests: []*workspaces.TerminateRequest{
+			{
+				WorkspaceId: aws.String(id),
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Waiting for workspace %q to be terminated", id)
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			workspaces.WorkspaceStatePending,
+			workspaces.WorkspaceStateAvailable,
+			workspaces.WorkspaceStateImpaired,
+			workspaces.WorkspaceStateUnhealthy,
+			workspaces.WorkspaceStateRebooting,
+			workspaces.WorkspaceStateStarting,
+			workspaces.WorkspaceStateRebuilding,
+			workspaces.WorkspaceStateRestoring,
+			workspaces.WorkspaceStateMaintenance,
+			workspaces.WorkspaceStateAdminMaintenance,
+			workspaces.WorkspaceStateSuspended,
+			workspaces.WorkspaceStateUpdating,
+			workspaces.WorkspaceStateStopping,
+			workspaces.WorkspaceStateStopped,
+			workspaces.WorkspaceStateError,
+		},
+		Target: []string{
+			workspaces.WorkspaceStateTerminating,
+			workspaces.WorkspaceStateTerminated,
+		},
+		Refresh:      workspaceRefreshStateFunc(conn, id),
+		PollInterval: 15 * time.Second,
+		Timeout:      10 * time.Minute,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("workspace %q was not terminated: %s", id, err)
+	}
+	log.Printf("[DEBUG] Workspace %q is terminated", id)
+
+	return nil
+}
+
+func workspaceRefreshStateFunc(conn *workspaces.WorkSpaces, workspaceID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := conn.DescribeWorkspaces(&workspaces.DescribeWorkspacesInput{
+			WorkspaceIds: []*string{aws.String(workspaceID)},
+		})
+		if err != nil {
+			return nil, workspaces.WorkspaceStateError, err
+		}
+		workspace := resp.Workspaces[0]
```

Comment:
Will probably want to check `len(resp.Workspaces) > 0` here and call `workspaceRefreshStateFunc()` in `resourceAwsWorkspaceRead()`.
See https://github.com/terraform-providers/terraform-provider-aws/pull/11837.

Reply:
Fixed